[Chrome] More Help Menu extensions with semantic links

cchaos opened the pull request:
## Consistency!

While the original help menu extension provided a much needed place to add individual docs and feedback links, there was a tragic amount of inconsistency in the way these links were presented. Here were all the different permutations:

**Before**

<img width="1916" alt="Screen Shot 2019-11-14 at 17 01 04 PM" src="https://user-images.githubusercontent.com/549577/68899764-5c68fc00-0700-11ea-9311-32efadd0f4f9.png">

**Now**

This PR updates the `chrome.helpExtension` that allows passing an array of different **link types** which will then spit out the correct form of the link so that they are consistent across all apps.

<img width="1651" alt="Screen Shot 2019-11-14 at 17 06 14 PM" src="https://user-images.githubusercontent.com/549577/68900091-18c2c200-0701-11ea-9620-2bbe9bcb5850.png">


### Updates to the API

The `chrome.helpExtension` has been updated to where it no longer just accepts a function to spit out any content. Now, the extension looks like:

```tsx
export interface ChromeHelpExtension {
  appName: string;
  links?: HeaderHelpMenuUIExtraLink[];
  // You can still pass custom content via the `content` key
  content?: (element: HTMLDivElement) => () => void;
}
```

Where `HeaderHelpMenuUIExtraLink`s can be one of these:

```tsx
type GitHubLink = EuiButtonEmptyProps & {
  /**
   * Creates a link to a new github issue in the Kibana repo
   */
  linkType: 'github';
  /**
   * Include at least one app-specific label to be applied to the new github issue
   */
  labels: string[];
  /**
   * Provides initial text for the title of the issue
   */
  title?: string;
};

type DiscussLink = EuiButtonEmptyProps & {
  /**
   * Creates a generic give feedback link with comment icon
   */
  linkType: 'discuss';
  href: string;
};

type DocumentationLink = EuiButtonEmptyProps & {
  /**
   * Creates a deep-link to app-specific documentation
   */
  linkType: 'documentation';
  href: string;
};

type CustomLink = EuiButtonEmptyProps & {
  /**
   * Extend EuiButtonEmpty to provide extra functionality
   */
  linkType: 'custom';
  text: React.ReactNode;
};
```

This will create the new section with the divider and the app's name. Then follow that with the links.

The most interesting link is the Github type. This follows the [query parameters that GH has setup](https://help.github.com/en/github/managing-your-work-on-github/about-automation-for-issues-and-pull-requests-with-query-parameters) for automatically applying labels and/or a title.

For example, the Lens github link leads to the template like so:

<img width="1121" alt="Screen Shot 2019-11-14 at 17 13 05 PM" src="https://user-images.githubusercontent.com/549577/68900539-0eed8e80-0702-11ea-8e54-3b4874dc5493.png">

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- ~[ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- ~[ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~

# Dev Docs

## ChromeHelpExtension

The `chrome.helpExtension` has been updated to where it no longer just accepts a function to spit out any content. Now, the extension looks like:

```tsx
export interface ChromeHelpExtension {
  appName: string;
  links?: HeaderHelpMenuUIExtraLink[];
  // You can still pass custom content via the `content` key
  content?: (element: HTMLDivElement) => () => void;
}
```

Commit:
Update help menu util to allow common types of links

- Github, Docs, and custom

Commit:
Fix test?

Commit:
Updating API docs?

Commit:
Quick layout fix

Commit:
Kibana core apps updated (Visualize, Dashboard, Discover)

Commit:
Updated Canvas usage

Commit:
Updated Maps usage

Commit:
Added a `discuss` link type

Commit:
Updated Logs and Metrics usage

Commit:
Update APM usage

Commit:
Update Uptime usage

Commit:
Update SIEM usage

Commit:
Removing old Uptime file

Commit:
Update Lens usage

cchaos committed:
Merge branch 'master' into help-links

cchaos commented:
@elastic/kibana-platform Hopefully I updated all the right API docs etc...

@elastic/logs-metrics-ui The link from Metrics to discuss is broken and doesn't seem to be a real link

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9935/)
* Commit: 396547f02888a9050dcedbc7437de8346d2e3d31

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9932/)
* Commit: 49bda887dec18413c52145f7456b8df1e0e60c06

smith reviewed (approved):
Much nicer. APM stuff looks good to me.

Commit:
Update API doc

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10102/)
* Commit: 41b3ff501dd5830c937d577c18641e4b16ed12bf

Commit:
[Uptime] Remove test for file that no longer exists

cchaos commented:
@elastic/uptime I completely removed a test file since the file (kibana_global_help.tsx) it was testing not longer exists.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10139/)
* Commit: b3b75b19576e4663a73a1704611a66b792888407

Commit:
One more snap and i18n tokens

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10194/)
* Commit: 9e516f89c273819e7aac9555a2ad597720ca0e32

pgayvallet commented on src/core/public/chrome/chrome_service.tsx:
```diff
@@ -58,7 +59,11 @@ export interface ChromeBrand {
+export interface ChromeHelpExtension {
+  appName: string;
+  links?: HeaderHelpMenuUIExtraLink[];
+  content?: (element: HTMLDivElement) => () => void;
```
From usages, `links` seems like a replacement for `content`, so we may want to define them as mutually exclusive instead of being both optionals? Or is there use cases where a developer may want to use them both?

cchaos replied:
There are use cases for both. For instance, Canvas displays both a documentation link and a custom link that opens a flyout with keyboard shortcuts.
https://github.com/elastic/kibana/pull/50736/files#diff-4b49c85975d24172c5abc437d0114b5d

pgayvallet reviewed (commented):
Great consistency improvement and `chrome.helpExtension.set` usage simplification! Looking good for platform team. 

Just some minor points:

sqren commented on x-pack/legacy/plugins/apm/public/index.tsx:
```diff
@@ -22,16 +24,28 @@ import { KibanaCoreContextProvider } from '../../observability/public';
+    {
+      linkType: 'discuss',
+      href: 'https://discuss.elastic.co/c/apm',
+      rel: 'noopener'
```
Would be nice if this was automatically appended to outbound links. Although that's perhaps outside the scope of this PR

sqren replied:
... actually: links of type `discuss` will always be outbound, correct? In that case it should be added automatically, no?

cchaos replied:
Yeah I'm not really familiar with this property and only saw a few instance use it, so I wasn't sure if necessary. I'll look up what it does to see if it makes sense to add automatically.

cchaos replied:
Updated to always use `noopener` on external links in help menu

sqren commented on x-pack/legacy/plugins/apm/public/components/app/GlobalHelpExtension/index.tsx:
```diff
@@ -4,46 +4,27 @@
+        {i18n.translate('xpack.apm.helpMenu.upgradeAssistantLink', {
+          defaultMessage: 'Upgrade assistant'
+        })}
+      </EuiLink>
```
Instead of rendering this via `content` I think it should be possible to declare custom links if an applicable type doesn't exist:

```js
{
  linkType: 'custom',
  icon: 'any valid eui icon',
  href: '/app/kibana/management/elasticsearch/upgrade_assistant',
  label: {i18n.translate('xpack.apm.helpMenu.upgradeAssistantLink', {
    defaultMessage: 'Upgrade assistant'
  })}
}
```

To me it seems like `content` should be a very last resort (it's exactly what'll cause inconsistencies again down the line), so giving consumers an option to avoid it would be nice.

sqren replied:
That being said, I think `upgrade_assistant` should be a dedicated link type.

cchaos replied:
I can certainly return an object instead of React content for this particular link. However, I'm going to leave the decision to make the upgrade assistant link a known link for later discussion. I only see this one usage and there may be a better way to use the Kibana Management page for this type of messaging.

cchaos replied:
Updated

sorenlouv reviewed (commented):
Awesome job @cchaos! Just a few minor comments.

weltenwort commented on x-pack/legacy/plugins/infra/public/pages/infrastructure/index.tsx:
```diff
@@ -41,10 +41,9 @@ export const InfrastructurePage = injectUICapabilities(
 
         <HelpCenterContent
           feedbackLink="https://discuss.elastic.co/c/metrics"
```
Thanks for catching this, we'll fix that separately (#50905).

weltenwort reviewed (approved):
nice improvement :heart:

`infra` changes LGTM

Commit:
PR feedback

- Wrap feedbackText in method
- Use `rel: noopener` on external links
- Rename HeaderHelpMenuUIExtraLink -> ChromeHelpExtensionMenuExtraLink and export

Commit:
[APM] Use ‘custom’ style link for Upgrade assistant

Commit:
API docs

cchaos commented on src/core/public/chrome/ui/header/header_help_menu.tsx:
```diff
@@ -137,6 +215,70 @@ class HeaderHelpMenuUI extends Component<Props, State> {
+          defaultMessage: 'Give feedback on {appName}',
+          values: { appName: helpExtension.appName },
+        }
+      );
```
Updated

cchaos commented on src/core/public/public.api.md:
```diff
@@ -119,7 +121,16 @@ export interface ChromeDocTitle {
+    // Warning: (ae-forgotten-export) The symbol "HeaderHelpMenuUIExtraLink" needs to be exported by the entry point index.d.ts
+    // 
+    // (undocumented)
+    links?: HeaderHelpMenuUIExtraLink[];
```
Renamed and exported

cchaos commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'master' into help-links

sqren commented on x-pack/legacy/plugins/apm/public/index.tsx:
```diff
@@ -4,34 +4,42 @@
+      linkType: 'custom',
+      href: url.format({
+        pathname: core.http.basePath.prepend('/app/kibana')
+      }),
```
Shouldn't this be:
```js
href={url.format({	
  pathname: core.http.basePath.prepend('/app/kibana'),	
  hash: '/management/elasticsearch/upgrade_assistant'	
})}
```
?

cchaos replied:
Thanks, must have been removed by accident. Fixed

Commit:
i18n

Commit:
Fix link

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10527/)
* Commit: d7baf50381b571151d9290df9ca9c73ca4734245

sorenlouv reviewed (approved):
LGTM, thanks for doing this 👍

pgayvallet reviewed (approved):
platform changes LGTM

cchaos commented:
@elastic/kibana-canvas @elastic/kibana-app Care to review your help links?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10542/)
* Commit: 482a35093a1fa84a953e6813bf6b3f4517ff6f68

cchaos committed:
Merge branch 'master' into help-links

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10671/)
* Commit: f2533a87fafa82083203fd5f390effc2314ecff7

cchaos commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'master' into help-links

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10739/)
* Commit: 86e4058dfb1b2699ef39d5d3b21ffa47321487d6

ryankeairns reviewed (approved):
👍 Canvas looks good, thank you. Link opens to correct docs page; keyboard shortcuts open in flyout.

Commit:
Merge remote-tracking branch 'upstream/master' into help-links

# Conflicts:
#	src/legacy/core_plugins/kibana/public/dashboard/help_menu/help_menu.js
#	src/legacy/core_plugins/kibana/public/dashboard/help_menu/help_menu_util.js
#	x-pack/legacy/plugins/lens/public/help_menu_util.tsx

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10956/)
* Commit: b8f7ad39db5c9294be431dc18bd3fc4cd52322d9

myasonik commented on src/core/public/chrome/ui/header/header_help_menu.tsx:
```diff
@@ -137,6 +215,74 @@ class HeaderHelpMenuUI extends Component<Props, State> {
+      customContent = (
+        <>
+          <EuiTitle size="xxs">
+            <span>{appName}</span>
```
Thoughts on changing this `<span>` to an `<h3>` and "Help" on line 314 (which I can't comment on) to an `h2`?
```suggestion
            <h3>{appName}</h3>
```

cchaos replied:
Sure if you think those heading levels are appropriate even when the content of the page is unknown

myasonik replied:
I think so because: 
- The `h1` should always be something about the page so we shouldn't be messing up the heading order
- I didn't want to skip any headings because, like you say, we don't know the rest of the page content
- These are custom, app specific help menus so I'm comfortable highlighting them to someone navigating by page headings
- Ultimately, these are actually pretty low risk/value either way because generally they won't even be rendered so it's mostly to let users quickly navigate this menu which isn't huge itself

So I think it's a nice little improvement without much risk but would also not fight too strongly for it if other people were opposed.

cchaos replied:
Thank you for the explanation! I have updated them to use h2 and h3.

myasonik commented on src/core/public/public.api.md:
```diff
@@ -119,7 +121,23 @@ export interface ChromeDocTitle {
 // @public (undocumented)
-export type ChromeHelpExtension = (element: HTMLDivElement) => () => void;
+export interface ChromeHelpExtension {
+    // (undocumented)
```
A lot of "Warning"s and "(undocumented)" bits in here and I'm not super certain what each means but it certainly seems scary. Can any of them be resolved?

cchaos replied:
Hmmm, let me see what happens if I export all those types

Commit:
Merge remote-tracking branch 'upstream/master' into help-links

Commit:
Documented ChromeHelpExtension props

Commit:
Use heading levels

Commit:
Update names of props, document prop, and update API docs

Commit:
Remove `ae-missing-release-tag` warnings

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11027/)
* Commit: d4aa3104c0e3377a6932c57a5938ae18e284ffec

cchaos commented:
Backport pending merge of https://github.com/elastic/kibana/pull/51204
